libsolidity/formal/CHC.h

```diff
@@ -0,0 +1,100 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/**
+ * Model checker based on Constrained Horn Clauses.
+ *
+ * A Solidity contract's CFG is encoded into a system of Horn clauses where
+ * each block has a predicate and edges are rules.
+ *
+ * The entry block is the constructor which has no in-edges.
+ * The constructor has one out-edge to an artificial block named _Interface_
+ * which has in/out-edges from/to all public functions.
+ *
+ * Loop invariants for Interface -> Interface' are state invariants.
+ */
+
+#pragma once
+
+#include <libsolidity/formal/SMTEncoder.h>
+
+#include <libsolidity/formal/CHCSolverInterface.h>
+
+#include <set>
+
+namespace dev
+{
+namespace solidity
+{
+
+class CHC: public SMTEncoder
+{
+public:
+	CHC(smt::EncodingContext& _context, langutil::ErrorReporter& _errorReporter);
+
+	void analyze(SourceUnit const& _sources, std::shared_ptr<langutil::Scanner> const& _scanner);
+
+	std::set<Expression const*> safeAssertions() { return m_safeAssertions; }
+
+private:
+	/// Visitor functions.
+	//@{
+	bool visit(ContractDefinition const& _node) override;
+	void endVisit(ContractDefinition const& _node) override;
+	bool visit(FunctionDefinition const& _node) override;
+	void endVisit(FunctionDefinition const& _node) override;
+	bool visit(IfStatement const& _node) override;
+	void endVisit(FunctionCall const& _node) override;
+
+	void visitAssert(FunctionCall const& _funCall);
+	//@}
+
+	/// Helpers.
+	//@{
+	void reset();
+	bool shouldVisit(ContractDefinition const& _contract);
+	bool shouldVisit(FunctionDefinition const& _function);
+	//@}
+
+	/// Solver related.
+	//@{
+	/// @returns true if query is unsatisfiable (safe).
+	bool query(smt::Expression const& _query, langutil::SourceLocation const& _location);
+	//@}
+
+	/// Verification targets.
+	//@{
+	std::vector<Expression const*> m_verificationTargets;
+
```

Comment:
Shouldn't these be seperated by another pair of `//@}` and `//@{`. Doxygen would group them together otherwise, wouldn't it? Or are they meant to be grouped?

Reply:
I meant them to be together because they're both related to verification targets.